src/pyrokinetics/local_geometry/local_geometry.py

```diff
@@ -1,152 +1,341 @@
+    @property
+    def _shape_params(self) -> ShapeParams:
+        data = {
+            field.name: getattr(self, field.name)
+            for field in dataclasses.fields(self.ShapeParams)
+        }
+        return self.ShapeParams(**data)
 
-    def __setattr__(self, key, value):
-        if value is None:
-            super().__setattr__(key, value)
+    @_shape_params.setter
+    def _shape_params(self, params: ShapeParams) -> None:
+        for field in dataclasses.fields(self.ShapeParams):
+            setattr(self, field.name, getattr(params, field.name))
 
-        if hasattr(self, key):
-            attr = getattr(self, key)
-            if hasattr(attr, "units") and not hasattr(value, "units"):
-                value *= attr.units
-                if not self._already_warned and str(attr.units) != "dimensionless":
-                    warn(
-                        f"missing unit from {key}, adding {attr.units}. To suppress this warning, specify units. Will "
-                        f"maintain units if not specified from now on"
-                    )
-                    self._already_warned = True
-        super().__setattr__(key, value)
-
-    def keys(self):
-        return self.__dict__.keys()
+    def _init_with_shape_params(
+        self,
+        psi_n: Float = DEFAULT_INPUTS["psi_n"],
+        rho: Float = DEFAULT_INPUTS["rho"],
+        Rmaj: Float = DEFAULT_INPUTS["Rmaj"],
+        Z0: Float = DEFAULT_INPUTS["Z0"],
+        a_minor: Float = DEFAULT_INPUTS["a_minor"],
+        Fpsi: Float = DEFAULT_INPUTS["Fpsi"],
+        FF_prime: Float = DEFAULT_INPUTS["FF_prime"],
+        B0: Float = DEFAULT_INPUTS["B0"],
+        q: Float = DEFAULT_INPUTS["q"],
+        shat: Float = DEFAULT_INPUTS["shat"],
+        beta_prime: Float = DEFAULT_INPUTS["beta_prime"],
+        dpsidr: Float = DEFAULT_INPUTS["dpsidr"],
+        bt_ccw: int = DEFAULT_INPUTS["bt_ccw"],
+        ip_ccw: int = DEFAULT_INPUTS["ip_ccw"],
+        theta: Optional[Array] = None,
+        overwrite_dpsidr: bool = True,
+        **shape_params,
+    ) -> None:
+        """Initialise a new instance using a given set of shaping parameters.
+
+        Used in the ``__init__`` functions of subclasses and when building from
+        a global equilibrium or another local geometry.
+
+        When building from GK input data, should always overwrite ``dpsidr``.
+        When building from equilibrium data or another local geometry, should not
+        overwrite ``dpsidr``.
+        """
+        if theta is None:
+            theta = np.linspace(0, 2 * np.pi, 256)
+        params = self.ShapeParams(**shape_params)
+
+        # Get flux surface curve, B_poloidal, and derivatives
+        R, Z = self._flux_surface(theta, Rmaj, Z0, rho, params)
+        derivatives = self._RZ_derivatives(theta, rho, params)
+        bunit_over_b0 = self._bunit_over_b0(Rmaj, Z0, rho, params)
+        if overwrite_dpsidr:
+            dpsidr = rho * bunit_over_b0 / q
```

Comment:
When initialising a local geometry, the code was previously overwriting `dpsidr` if it was provided via a GK input file, but not if it was determined via an equilibrium or another local geometry. This feels like a bug.

This check is included as otherwise a few tests fail that I can't verify for myself. It also necessitates an `overwrite_dpsidr` variable in all of the subclass `__init__` functions. If this is unintended behaviour, I can remove it throughout.

Reply:
`dpsidr` should be determined from `LocalGeometry` parameters alone, you never actually specify it as an input to a code. You can of course get it from a full equilibrium so that could be a good check but in theory you should always take the one from the `LocalGeometry` parameters, its like a fit I guess.  I guess it should always be overwritten.

Reply:
I've been playing around with this and I ran into a bit of a units issue. If we build from a GK input file, then everything is just a `float` until we attach units during the normalisation step, after which `dpsidr` is assigned units of `lref * bref`. However, if we calculate it from an equilibrium file, `dpsidr` will have units of just `[length]`, as `rho * bunit_over_b0 / q` has units `[length] [dimensionless] / [dimensionless]`, and when we try to normalise this to `lref * bref` it throws an error. I think we need to use the provided `dpsidr` when building from an equilibrium, but otherwise should be fine to recalculate.

Reply:
I did some further playing around and it seems this issue appears when building from another local geometry too. It looks like the best solution is to use `dpsidr` if it's available, and only calculate it as a backup. I've come up with a cleaner solution to the problem though, so it's not a big deal -- not sure why I thought I needed `overwrite_dpsidr`, as it seems just using a default value of `None` is sufficient.

Reply:
The issue here with `bunit_over_b0` is that it technically is `Bunit` with units of `B0` so when `bref` = `B0` then it dimensionally makes sense. To be honest it may be worth adding in the correct units for `Bunit` which is a just a magnetic  field rather than having `Bunit/B0`.